google/cloud/security/common/data_access/group_dao.py

```diff
@@ -25,28 +25,68 @@
+    def get_group_id(self, resource_name, email, timestamp):
+        """Get the id of a group from an email.
+
+        Args:
+            resource_name: String of the resource name.
+            group_id: String of the group id.
+            timestamp: The timestamp of the snapshot.
+
+        Returns:
+             String of the group id.
+        """
+        sql = select_data.GROUP_ID.format(email, timestamp)
+        result = self.execute_sql_with_fetch(resource_name, sql, None)
+        return result[0].get('group_id')
+
+    def get_group_members(self, resource_name, group_id, timestamp):
+        """Get the members of a group.
 
         Args:
+            resource_name: String of the resource name.
+            group_id: String of the group id.
             timestamp: The timestamp of the snapshot.
 
         Returns:
-             A dict containing the group (gcp_type.group) and a list of
-            their members (gcp_type.group_members) who are user type.
+             A tuple of group members in dict format.
+             ({'group_id': '00lnxb',
+               'member_email': 'foo@mygbiz.com',
+               'member_id': '11111',
+               'member_role': 'OWNER',
+               'member_type': 'USER'}, ...)
         """
-        sql = select_data.GROUP_USERS.format(timestamp)
-        rows = self.execute_sql_with_fetch(resource_name, sql, None)
-        group_users = {}
-        for row in rows:
-            # TODO: Determine if parenting is needed here, similar to
-            # project IAM policies.
-            group_user = group_member.GroupMember(
-                row.get('member_role'),
-                row.get('member_type'),
-                row.get('member_email'))
-            if row.get('group_id') not in group_users:
-                group_users[row.get('group_id')] = [group_user]
-            else:
-                group_users[row.get('group_id')].append(group_user)
-        return group_users
+        sql = select_data.GROUP_MEMBERS.format(group_id, timestamp)
+        return self.execute_sql_with_fetch(resource_name, sql, None)
+
+    def get_recursive_members_of_group(self, group_email, timestamp):
+        """Get all the recursive members of a group.
+
+        Args:
+            group_email: String of the group email.
+            timestamp: The timestamp of the snapshot.
+
+        Returns:
+             A list of group members in dict format.
+             [{'group_id': '00lnxb',
+               'member_email': 'foo@mygbiz.com',
+               'member_id': '11111',
+               'member_role': 'OWNER',
+               'member_type': 'USER'}, ...]
+        """
+        all_members = []
+        queue = Queue()
+
+        group_id = self.get_group_id(
+            'group', group_email, timestamp)
+        queue.put(group_id)
+
+        while not queue.empty():
+            group_id = queue.get()
+            members = self.get_group_members('group_members', group_id,
```

Comment:
nit: 'group_members' can we reference this type somewhere that's statically defined?